casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -58,19 +60,29 @@ object Running {
             casper => {
               for {
                 latestMessages <- casper.blockDag.flatMap(
-                                   _.latestMessageHashes.map(_.values.toList)
+                                   _.latestMessageHashes.map(_.values.toSet)
                                  )
-                tips            <- latestMessages.traverse(BlockStore[F].getUnsafe)
-                newestTimestamp = tips.map(_.header.timestamp).max
-                now             <- Time[F].currentMillis
-                expired         = (now - newestTimestamp) > delayThreshold.toMillis
+                now <- Time[F].currentMillis
+                hasRecentLatestMessage = Stream
+                  .fromIterator(latestMessages.iterator)
+                  .evalMap(BlockStore[F].getUnsafe)
+                  // filter only blocks that are recent
+                  .filter { b =>
+                    val blockTimestamp = b.header.timestamp
+                    (now - blockTimestamp) < delayThreshold.toMillis
+                  }
+                  .head
+                  .compile
+                  .last
```

Comment:
`head` and then `last`. Hm...I don't actually get the better idea. I think I understand your intention. You just want to get one of the result.

Reply:
Not a big deal